test/python/WMCore_t/MicroService_t/Unified_t/MSRuleCleaner_t.py

```diff
@@ -0,0 +1,197 @@
+        self.msRuleCleaner.plineArchive.run(wflow)
+        expectedWflow = {'CleanupStatus': {'plineAgentBlock': True, 'plineAgentCont': True},
+                         'ForceArchive': False,
+                         'IsClean': True,
+                         'OutputDatasets': [u'/JetHT/CMSSW_7_2_0-RECODreHLT_TaskChain_LumiMask_multiRun_HG2011_Val_Todor_v1-v11/RECO',
+                                            u'/JetHT/CMSSW_7_2_0-RECODreHLT_TaskChain_LumiMask_multiRun_HG2011_Val_Todor_v1-v11/DQMIO',
+                                            u'/JetHT/CMSSW_7_2_0-SiStripCalMinBias-RECODreHLT_TaskChain_LumiMask_multiRun_HG2011_Val_Todor_v1-v11/ALCARECO',
+                                            u'/JetHT/CMSSW_7_2_0-SiStripCalZeroBias-RECODreHLT_TaskChain_LumiMask_multiRun_HG2011_Val_Todor_v1-v11/ALCARECO',
+                                            u'/JetHT/CMSSW_7_2_0-TkAlMinBias-RECODreHLT_TaskChain_LumiMask_multiRun_HG2011_Val_Todor_v1-v11/ALCARECO'],
+                         'ParentageResolved': True,
+                         'PlineMarkers': ['plineArchive',
+                                          'plineAgentBlock',
+                                          'plineAgentCont',
+                                          'plineArchive'],
+                         'RequestName': u'TaskChain_LumiMask_multiRun_HG2011_Val_201029_112735_5891',
+                         'RequestStatus': u'announced',
+                         'RequestType': u'TaskChain',
+                         'RulesToClean': {'plineAgentBlock': [], 'plineAgentCont': []},
+                         'TargetStatus': None,
+                         'TransferDone': False}
+        self.assertDictEqual(wflow, expectedWflow)
+
+        # Try archival of an uncleaned workflow
+        wflow['CleanupStatus']['plineAgentBlock'] = False
+        with self.assertRaises(MSRuleCleanerArchival):
+            self.msRuleCleaner.plineArchive.run(wflow)
+
+    def testRunning(self):
+        result = self.msRuleCleaner._execute(self.reqRecords)
+        self.assertEqual(result, (3, 2, 2))
+
+    def testCheckClean(self):
+        # NOTE: All of the bellow checks are well visualized at:
+        #       https://github.com/dmwm/WMCore/pull/10023#discussion_r520070925
+
+        # 1. MaskList shorter than FlagList
+        wflowFlags = {'CleanupStatus': {'plineAgentBlock': True, 'plineAgentCont': True, 'plineMStrCont': False},
+                      'PlineMarkers': ['plineAgentBlock', 'plineAgentCont']}
+        self.assertTrue(self.msRuleCleaner._checkClean(wflowFlags))
+
+        wflowFlags = {'CleanupStatus': {'plineAgentBlock': False, 'plineAgentCont': True, 'plineMStrCont': True},
+                      'PlineMarkers': ['plineAgentBlock', 'plineAgentCont']}
+        self.assertFalse(self.msRuleCleaner._checkClean(wflowFlags))
+
+        # 2. MaskList Empty
+        wflowFlags = {'CleanupStatus': {'plineAgentBlock': True, 'plineAgentCont': True},
```

Comment:
Just out of curiosity, is it a valid case? I mean, can we have a pipeline name in the CleanupStatus dict but not marked in the PlineMarkers?

Reply:
No, The code as  it is right now should have other protections against this case, but I put this check the unit test for completeness, because the `_checkClean()` function mast cover all the 4 types  of combinations regardless if the rest of the code has an additional protection or not. In this case - the `cleanRucioRules()` function is reading the `currentPline` from the `PlineMarkers` list, but in case we make a change about this, I think this should have been thought of in advance in the `_checkClean()` and we assure the code is doing the right thing even if we try to break it.